### PR TITLE
[FIX JENKINS-43228] Consider time zone for cron validation

### PR DESCRIPTION
**Pull request by daniel-beck**
# Description

See [JENKINS-43228](https://issues.jenkins-ci.org/browse/JENKINS-43228).

Details: Time zones defined for cron weren't considered for the form validation (previous/next occurrence) 

**Untested, waiting for PR build for test results**

### Changelog entries

Proposed changelog entries:

* Consider time zone for cron validation.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Link to JIRA ticket in description, if appropriate
- [x] Appropriate autotests or explanation to why this change has no tests
- [n/a] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

### Desired reviewers

@oleg-nenashev who still hasn't finished #1720. We're soon celebrating it's second birthday ;-)

### CODE REVIEW

**Comment by jglick** [comment]
> No test as this concerns time zone difference from current time zone, and I'm not sure how to test that without it being terrible or having side effects for other tests.

IIRC there are already some tests which set the default timezone and restore it in a `finally`-block. Since each test JVM runs only one test case at a given time, this is safe enough.

File: core/src/main/java/hudson/scheduler/CronTabList.java [hunk 1]
```diff
@@ -131,7 +131,7 @@ public static CronTabList create(@Nonnull String format, Hash hash) throws ANTLR
     public @CheckForNull Calendar previous() {
         Calendar nearest = null;
         for (CronTab tab : tabs) {
-            Calendar scheduled = tab.floor(Calendar.getInstance());
+            Calendar scheduled = tab.floor(tab.getTimeZone() == null ? Calendar.getInstance() : Calendar.getInstance(tab.getTimeZone()));
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scheduler/CronTabList.java]: FindBugs will likely gruble about that. Better to use a temporary variable for `tab.getTimeZone()`

File: core/src/main/java/hudson/scheduler/CronTab.java [hunk 2]
```diff
@@ -532,4 +532,17 @@ private String toString(String key, long bit) {
...
         }
     }
+
+    /**
+     * Returns the configured time zone, or null if none is configured
+     *
+     * @return the configured time zone, or null if none is configured
+     * @since TODO
+     */
+    @CheckForNull public TimeZone getTimeZone() {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/scheduler/CronTab.java]: If the fix is a subject for backporting, better to mark this API method as restricted for now

Comment (daniel-beck) [inline core/src/main/java/hudson/scheduler/CronTab.java]: That can always be done during backporting. There's no reason it should be restricted here.

File: core/src/main/java/hudson/scheduler/CronTabList.java (hunk 1 above)
Comment (daniel-beck) [inline core/src/main/java/hudson/scheduler/CronTabList.java]: PR build passed, and while it's not enabled, I see no reason to placate it.

Comment (oleg-nenashev) [inline core/src/main/java/hudson/scheduler/CronTabList.java]: OK in thus case (no real risk of the issue)

**Comment by daniel-beck** [comment]
@oleg-nenashev As I mentioned yesterday, this isn't approved by you yet.

**Review by oleg-nenashev — APPROVED** [review]
@daniel-beck approved, sry

**Pull request merged**